### PR TITLE
Add first version of external API documentation

### PR DESCRIPTION
**Pull request by martijnrusschen**
Ref #291 

First step in publishing a technical API in on our GH pages website. This PR focusses on generating useful output.

Output looks like https://github.com/Hacker0x01/react-datepicker/blob/api-docs/docs/datepicker.md

### CODE REVIEW

File: src/datepicker.jsx [hunk 1]
```diff
@@ -108,7 +107,7 @@ var DatePicker = React.createClass({
       ref="calendar"
       weekdays={this.props.weekdays}
       locale={this.props.locale}
-      moment={this.props.moment}
```
Comment (martijnrusschen) [inline src/datepicker.jsx]: Need to investigate how I can get this working with moment as defaultProp

File: docs/src/root.jsx [hunk 2]
```diff
@@ -1,8 +1,10 @@
 import React from "react";
+import { Router, Route, IndexRoute, Link, browserHistory } from "react-router";
```
Comment (martijnrusschen) [inline docs/src/root.jsx]: These changes shouldn't be part of this PR. Focussing on generating useful output first.

File: package.json [hunk 3]
```diff
@@ -77,7 +77,9 @@
   "dependencies": {
     "classnames": "^2.2.1",
     "moment": "^2.11.2",
+    "react-docgen": "^2.7.0",
     "react-onclickoutside": "^4.5.0",
+    "react-router": "^2.0.0",
```
Comment (martijnrusschen) [inline package.json]: Revert

File: src/datepicker.jsx (hunk 1 above)
Comment (martijnrusschen) [inline src/datepicker.jsx]: Error:

```
Error with path "src/datepicker.jsx": Error: [{type: ImportDefaultSpecifier, start: 7, end: 13, loc: [object Object], local: [object Object], id: null, name: null}] does not match type Printable
Error: [{type: ImportDefaultSpecifier, start: 7, end: 13, loc: [object Object], local: [object Object], id: null, name: null}] does not match type Printable
    at Type.Tp.assert (/Users/martijn/.nvm/versions/node/v4.2.3/lib/node_modules/react-docgen/node_modules/recast/node_modules/ast-types/lib/types.js:56:15)
    at genericPrintNoParens (/Users/martijn/.nvm/versions/node/v4.2.3/lib/node_modules/react-docgen/node_modules/recast/lib/printer.js:219:26)
    at genericPrint (/Users/martijn/.nvm/versions/node/v4.2.3/lib/node_modules/react-docgen/node_modules/recast/lib/printer.js:164:9)
    at printRootGenerically (/Users/martijn/.nvm/versions/node/v4.2.3/lib/node_modules/react-docgen/node_modules/recast/lib/printer.js:105:15)
    at maybeReprint (/Users/martijn/.nvm/versions/node/v4.2.3/lib/node_modules/react-docgen/node_modules/recast/lib/printer.js:97:16)
    at print (/Users/martijn/.nvm/versions/node/v4.2.3/lib/node_modules/react-docgen/node_modules/recast/lib/printer.js:75:29)
    at exports.printComments (/Users/martijn/.nvm/versions/node/v4.2.3/lib/node_modules/react-docgen/node_modules/recast/lib/comments.js:324:22)
    at printWithComments (/Users/martijn/.nvm/versions/node/v4.2.3/lib/node_modules/react-docgen/node_modules/recast/lib/printer.js:61:16)
    at print (/Users/martijn/.nvm/versions/node/v4.2.3/lib/node_modules/react-docgen/node_modules/recast/lib/printer.js:66:20)
    at Printer.print (/Users/martijn/.nvm/versions/node/v4.2.3/lib/node_modules/react-docgen/node_modules/recast/lib/printer.js:118:21)
```

Comment (rafeememon) [inline src/datepicker.jsx]: Probably should just remove moment as a prop entirely and just import.  I've been meaning to do that for a while, ha

**Comment by rafeememon** [comment]
Looks like some good stuff.  As a side note, I'm not sure we should document the API of _internal_ components, e.g. `Day`.  If we document it then we need to support it, and having the freedom to change it would be great.

**Comment by martijnrusschen** [comment]
What about just documenting the datepicker.jsx? Or do you see any other files that could be useful to document?

File: src/datepicker.jsx (hunk 1 above)
Comment (martijnrusschen) [inline src/datepicker.jsx]: Okay with this change? We could move the moment to the calendar.jsx in a different PR. (requires too much changes to keep this clean). Is there any use case where moment as a prop is useful?

**Comment by rafeememon** [comment]
@martijnrusschen, probably just `DatePicker` right now, but we'll add more when we start making other components as public API, like `Calendar` and `DateInput`

**Comment by martijnrusschen** [comment]
Sounds good, wrapping up this initial PR with just `Datepicker.jsx`

File: docs/generateMarkdown.sh [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+function generateProps(props) {
+  var title = 'Props';
+
+  return (
+    title + '\n' +
+    stringOfLength('-', title.length) + '\n' +
+    '\n' +
+    Object.keys(props).sort().map(function(propName) {
```
Comment (martijnrusschen) [inline docs/generateMarkdown.sh]: Bash always fun:
````
➜  react-datepicker git:(api-docs) react-docgen src/datepicker.jsx | docs/buildDocs.sh
/Users/martijn/Development/react-datepicker/docs/generateMarkdown.sh:65
    Object.keys(props).sort().map(function(propName) {
           ^

TypeError: Cannot convert undefined or null to object
    at Function.keys (native)
    at generateProps (/Users/martijn/Development/react-datepicker/docs/generateMarkdown.sh:65:12)
    at generateMarkdown (/Users/martijn/Development/react-datepicker/docs/generateMarkdown.sh:75:5)
    at buildDocs (/Users/martijn/Development/react-datepicker/docs/buildDocs.sh:29:20)
    at Socket.<anonymous> (/Users/martijn/Development/react-datepicker/docs/buildDocs.sh:22:3)
    at emitNone (events.js:72:20)
    at Socket.emit (events.js:166:7)
    at endReadableNT (_stream_readable.js:905:12)
    at doNTCallback2 (node.js:441:9)
    at process._tickCallback (node.js:355:17)
```

Generating all files works as a charm, but only generating datepicker.jsx errors.

Comment (martijnrusschen) [inline docs/generateMarkdown.sh]: Workaround, generate all files and exclude the ones you don't need :)
